File: Earth_Library/scripts/get_related.py
from __future__ import annotations

import argparse
import json
import os
from pathlib import Path

from el_parsers import load_jsonl, parse_tags

ROOT = Path(os.environ.get("CURSOR_PROJECT_DIR", Path(__file__).resolve().parents[2]))
CARDS_JSONL = ROOT / "Earth_Library" / "cards.jsonl"
RELATIONS_JSONL = ROOT / "Earth_Library" / "relations.jsonl"


def main() -> None:
    parser = argparse.ArgumentParser(description="Earth Library 关联推荐")
    parser.add_argument("--card_id", required=True, help="目标卡片 ID（如 20260508-205830_公司生命周期）")
    parser.add_argument("--relation_types", default="标签相交,关键词相交", help="关联类型，逗号分隔")
    parser.add_argument("--max_results", type=int, default=5, help="最大结果数")
    args = parser.parse_args()

    rel_types = {t.strip() for t in args.relation_types.split(",") if t.strip()}
    cards = load_jsonl(CARDS_JSONL)
    relations = load_jsonl(RELATIONS_JSONL)

    # 构建 id → card 映射
    card_map: dict[str, dict] = {}
    for c in cards:
        cid = c.get("id", "")
        if cid:
            card_map[cid] = c

    # 按关联类型筛选
    related_ids: set[str] = set()
    target_id = args.card_id
    for rel in relations:
        r_type = rel.get("r", "")
        if r_type not in rel_types:
            continue
        s = rel.get("s", "")
        t = rel.get("t", "")
        if s == target_id:
            related_ids.add(t)
        elif t == target_id:
            related_ids.add(s)

    # 构建结果
    hits: list[dict] = []
    for rid in list(related_ids)[:args.max_results]:
        card = card_map.get(rid)
        if not card:
            continue
        hits.append({
            "id": rid,
            "title": card.get("title", ""),
            "type": card.get("type", ""),
            "confidence": card.get("confidence", ""),
            "domain": card.get("domain", ""),
            "tags": parse_tags(card.get("tags", [])),
        })

    print(json.dumps({
        "ok": True,
        "card_id": target_id,
        "related": hits,
        "total": len(hits),
    }, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: Earth_Library/scripts/get_related.mjs
import { readFileSync, statSync } from "node:fs";
import { join } from "node:path";

const root = process.env.CURSOR_PROJECT_DIR || process.cwd();
const cardsPath = join(root, "Earth_Library", "cards.jsonl");
const relPath = join(root, "Earth_Library", "relations.jsonl");

const args = Object.fromEntries(
  process.argv.slice(2).map((v, i, a) => (v.startsWith("--") ? [v.slice(2), a[i + 1]] : null)).filter(Boolean)
);

const cardId = String(args.card_id || "").trim();
if (!cardId) {
  process.stdout.write(JSON.stringify({ ok: false, error: "missing --card_id" }));
  process.exit(1);
}

const relTypes = new Set(String(args.relation_types || "标签相交,关键词相交").split(",").map((s) => s.trim()));
const maxResults = parseInt(args.max_results || "5", 10);

function parseTags(raw) {
  if (typeof raw === "string") return raw.replace(/，/g, ",").split(",").map((s) => s.trim()).filter(Boolean);
  if (Array.isArray(raw)) return raw.flatMap((item) => parseTags(item));
  return [];
}

// Load cards map
const cardMap = {};
if (statSync(cardsPath, { throwIfNoEntry: false })?.isFile()) {
  for (const line of readFileSync(cardsPath, "utf8").split("\n")) {
    if (!line.trim()) continue;
    try {
      const c = JSON.parse(line);
      if (c.id) cardMap[c.id] = c;
    } catch {}
  }
}

// Load relations
const relatedIds = new Set();
if (statSync(relPath, { throwIfNoEntry: false })?.isFile()) {
  for (const line of readFileSync(relPath, "utf8").split("\n")) {
    if (!line.trim()) continue;
    try {
      const rel = JSON.parse(line);
      if (!relTypes.has(rel.r)) continue;
      if (rel.s === cardId && rel.t) relatedIds.add(rel.t);
      else if (rel.t === cardId && rel.s) relatedIds.add(rel.s);
    } catch {}
  }
}

const hits = [...relatedIds].slice(0, maxResults)
  .map((rid) => {
    const c = cardMap[rid];
    if (!c) return null;
    return {
      id: rid,
      title: c.title || "",
      type: c.type || "",
      confidence: c.confidence || "",
      domain: c.domain || "",
      tags: parseTags(c.tags || []),
    };
  })
  .filter(Boolean);

process.stdout.write(JSON.stringify({ ok: true, card_id: cardId, related: hits, total: hits.length }));
